src/runtime/security/sasl_utils.cpp

```diff
@@ -0,0 +1,156 @@
+#include <dsn/utility/flags.h>
+#include <dsn/utility/synchronize.h>
+
+namespace dsn {
+namespace security {
+DSN_DEFINE_string("security", sasl_plugin_path, "/usr/lib/sasl2", "path to search sasl plugins");
+
+const char *logger_level_to_string(int level)
+{
+    switch (level) {
+    case SASL_LOG_NONE:
+        return "SASL_LOG_NONE";
+    case SASL_LOG_ERR:
+        return "SASL_LOG_ERR";
+    case SASL_LOG_FAIL:
+        return "SASL_LOG_FAIL";
+    case SASL_LOG_WARN:
+        return "SASL_LOG_WARN";
+    case SASL_LOG_NOTE:
+        return "SASL_LOG_NOTE";
+    case SASL_LOG_DEBUG:
+        return "SASL_LOG_DEBUG";
+    case SASL_LOG_TRACE:
+        return "SASL_LOG_TRACE";
+    case SASL_LOG_PASS:
+        return "SASL_LOG_PASS";
+    default:
+        return "Unkown SASL log level";
+    }
+}
+
+int sasl_simple_logger(void *context, int level, const char *msg)
+{
+    if (SASL_LOG_NONE == level || nullptr == msg) {
+        return SASL_OK;
+    }
+
+    ddebug_f("sasl log info: log level = {}, message = {}", logger_level_to_string(level), msg);
+    return SASL_OK;
+}
+
+int sasl_get_path(void *context, char **path)
+{
+    if (nullptr == path) {
+        return SASL_BADPARAM;
+    }
+    *path = const_cast<char *>(FLAGS_sasl_plugin_path);
+    return SASL_OK;
+}
+
+int sasl_get_username(void *context, int id, const char **result, unsigned *len)
+{
+    if (nullptr == result) {
+        return SASL_BADPARAM;
+    }
+    static const std::string username = get_username();
+    switch (id) {
+    case SASL_CB_USER:
+    case SASL_CB_AUTHNAME:
+        *result = username.c_str();
+        if (len != nullptr) {
+            *len = username.length();
+        }
+        return SASL_OK;
+    default:
+        dassert_f(false, "unexpected SASL callback type: {}", id);
+        return SASL_BADPARAM;
+    }
+}
+
+sasl_callback_t client_callbacks[] = {
+    {SASL_CB_USER, (sasl_callback_ft)&sasl_get_username, nullptr},
+    {SASL_CB_GETPATH, (sasl_callback_ft)&sasl_get_path, nullptr},
+    {SASL_CB_AUTHNAME, (sasl_callback_ft)&sasl_get_username, nullptr},
+    {SASL_CB_LOG, (sasl_callback_ft)&sasl_simple_logger, nullptr},
+    {SASL_CB_LIST_END, nullptr, nullptr}};
+
+sasl_callback_t server_callbacks[] = {{SASL_CB_LOG, (sasl_callback_ft)&sasl_simple_logger, nullptr},
+                                      {SASL_CB_GETPATH, (sasl_callback_ft)&sasl_get_path, nullptr},
+                                      {SASL_CB_LIST_END, nullptr, nullptr}};
+
+// provide mutex function for sasl
+void *sasl_mutex_alloc_local() { return static_cast<void *>(new utils::ex_lock_nr); }
+
+void sasl_mutex_free_local(void *m) { delete static_cast<utils::ex_lock_nr *>(m); }
+
+int sasl_mutex_lock_local(void *m)
+{
+    static_cast<utils::ex_lock_nr *>(m)->lock();
+    return 0;
+}
+
+int sasl_mutex_unlock_local(void *m)
+{
+    static_cast<utils::ex_lock_nr *>(m)->unlock();
+    return 0;
+}
+
+void sasl_set_mutex_local()
+{
+    sasl_set_mutex(&sasl_mutex_alloc_local,
```

Comment:
I don't find `sasl_set_mutex`, I guess it is a function of `sasl/sasl.h`, could you please add some comments for this function or tell us where to find this function?